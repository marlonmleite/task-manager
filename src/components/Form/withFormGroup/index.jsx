import React from 'react'
import { Field } from 'formik'
import FormGroup from '../FormGroup'

const FormGroupHoc = Input => (props) => {
  const { label, field, form, ...otherProps } = props
  const { touched, errors } = form
  const error = touched[field.name] && errors[field.name]

  return (
    <FormGroup label={label} name={field.name} error={error}>
      <Input {...field} {...otherProps} />
    </FormGroup>
  )
}

const withFormGroup = (Input) => {
  const FieldComponent = FormGroupHoc(Input)

  return props => <Field {...props} component={FieldComponent} />
}

export default withFormGroup
export { FormGroupHoc }
