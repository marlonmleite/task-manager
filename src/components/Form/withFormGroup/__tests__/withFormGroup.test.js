import React from 'react'
import { shallow } from 'enzyme'
import withFormGroup, { FormGroupHoc } from '../'

const CustomInput = props => <input {...props} />
const Input = FormGroupHoc(CustomInput)

describe('withFormGroup', () => {
  it('should render', () => {
    const InputHoc = withFormGroup(CustomInput)
    const input = shallow(<InputHoc />)

    expect(input).toMatchSnapshot()
  })

  it('should render input', () => {
    const form = { touched: {}, errors: {} }
    const props = { field: { name: 'fullName', value: 'hi' }, form }
    const input = shallow(<Input {...props} label="" />)
    const field = input.find(CustomInput)

    expect(field.prop('name')).toBe('fullName')
    expect(field.prop('value')).toBe('hi')
  })

  it('should render with label', () => {
    const form = { touched: {}, errors: {} }
    const props = { field: { name: 'a', value: 'b' }, form, label: 'Name' }
    const input = shallow(<Input {...props} />)
    const group = input.find('FormGroup')

    expect(group.prop('label')).toBe('Name')
  })

  it('should display error', () => {
    const form = { touched: { a: true }, errors: { a: 'Required' } }
    const props = { field: { name: 'a' }, form }
    const input = shallow(<Input {...props} label="" />)
    const group = input.find('FormGroup')

    expect(group.prop('error')).toBe('Required')
  })

  it('should handle onChange', () => {
    const onChange = jest.fn()
    const form = { touched: {}, errors: {} }
    const props = { field: { name: 'fullName', onChange }, form }
    const input = shallow(<Input {...props} label="" />)

    input.find(CustomInput).simulate('change', 'ðŸš€')

    expect(onChange).toHaveBeenCalledWith('ðŸš€')
  })
})

