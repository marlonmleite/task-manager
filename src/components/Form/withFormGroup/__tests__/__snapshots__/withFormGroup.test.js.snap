// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFormGroup should render 1`] = `
<C
  component={[Function]}
/>
`;
